### PR TITLE
Optimize LMD-GHOST Vote Counting

### PR DESCRIPTION
**Pull request by rauljordan**
Part of #1586 

---

# Description

**Write why you are making the changes in this pull request**

Our fork-choice algorithm is extremely slow, often on the order of 2 or 3 seconds.

**Write a summary of the changes you are making**

This PR optimizes LMD GHOST to deal with an attestation target data structure that is more lightweight than a block. It contains only the information LMD GHOST requires, which are hashes and slots of attestation targets, preventing repeated hashing or expensive operations such as `proto.Equal` when all we need is a basic comparison of values. 

**Link anything that would be helpful or relevant to the reviewers**

Diff is large due to proto generation.

### CODE REVIEW

File: beacon-chain/db/state.go [hunk 1]
```diff
@@ -88,6 +88,32 @@ func (db *BeaconDB) InitializeState(ctx context.Context, genesisTime uint64, dep
 	})
 }
 
+// RiskyHeadState --
```
Comment (terenc3t) [inline beacon-chain/db/state.go]: comment

File: beacon-chain/attestation/service.go [hunk 2]
```diff
@@ -92,7 +92,7 @@ func (a *Service) IncomingAttestationFeed() *event.Feed {
 //		Attestation` be the attestation with the highest slot number in `store`
 //		from the validator with the given `validator_index`
 func (a *Service) LatestAttestation(ctx context.Context, index uint64) (*pb.Attestation, error) {
-	state, err := a.beaconDB.HeadState(ctx)
+	state, err := a.beaconDB.RiskyHeadState(ctx)
```
Comment (nisdas) [inline beacon-chain/attestation/service.go]: revert this back to HeadState, we need our copy of state to be immutable

File: beacon-chain/blockchain/fork_choice.go [hunk 3]
```diff
@@ -181,7 +184,11 @@ func (c *ChainService) ApplyForkChoiceRule(
 	if err != nil {
 		return fmt.Errorf("could not hash head: %v", err)
 	}
-	log.WithField("headRoot", fmt.Sprintf("0x%x", h)).Info("Chain head block and state updated")
+	endTime := time.Now()
+	log.WithFields(logrus.Fields{
+		"headRoot": fmt.Sprintf("0x%x", h),
+		"time":     fmt.Sprintf("%v", endTime.Sub(startTime)),
```
Comment (nisdas) [inline beacon-chain/blockchain/fork_choice.go]: I think the time was only added for testing, we dont need this in runtime

File: beacon-chain/db/state.go [hunk 4]
```diff
@@ -88,6 +88,37 @@ func (db *BeaconDB) InitializeState(ctx context.Context, genesisTime uint64, dep
 	})
 }
 
+// RiskyHeadState accesses the underlying cached state without cloning it. It is dangerous to call this
+// method if you are doing any write operations to the beacon state.
+func (db *BeaconDB) RiskyHeadState(ctx context.Context) (*pb.BeaconState, error) {
```
Comment (nisdas) [inline beacon-chain/db/state.go]: Not sure if we should even be using this, this is unsafe by default. We would have to mentally make a note whenever we use this method that the cached state isnt being mutated by what we are doing with it

File: beacon-chain/blockchain/fork_choice.go [hunk 5]
```diff
@@ -363,53 +369,63 @@ func VoteCount(block *pb.BeaconBlock, state *pb.BeaconState, targets map[uint64]
...
+	newTarget := &pb.AttestationTarget{
+		Slot:       parent.Slot,
+		BlockRoot:  parentRoot[:],
+		ParentRoot: parent.ParentRootHash32,
+	}
+	return BlockAncestor(newTarget, slot, beaconDB)
 }
 
-// cachedAncestorBlock retrieves the cached ancestor block from block ancestor cache,
+// cachedAncestorCached retrieves the cached ancestor target from block ancestor cache,
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/fork_choice.go]: Rename?

File: beacon-chain/blockchain/fork_choice.go [hunk 6]
```diff
@@ -363,53 +369,63 @@ func VoteCount(block *pb.BeaconBlock, state *pb.BeaconState, targets map[uint64]
...
+		return targetBlock.BlockRoot[:], nil
 	}
-	if block.Slot < slot {
+	if targetBlock.Slot < slot {
 		return nil, nil
 	}
-	parentHash := bytesutil.ToBytes32(block.ParentRootHash32)
-	parent, err := beaconDB.Block(parentHash)
+	parentRoot := bytesutil.ToBytes32(targetBlock.ParentRoot)
+	parent, err := beaconDB.Block(parentRoot)
```
Comment (shayzluf) [inline beacon-chain/blockchain/fork_choice.go]: can we get parent from map instead of reading from db?

Comment (rauljordan) [reply beacon-chain/blockchain/fork_choice.go]: Ah unfortunately not in this function, as the ancestor lookup can be different than just fetching the parent

**Pull request merged**